pelican_url/query.go

```diff
@@ -0,0 +1,131 @@
+/***************************************************************
+*
+* Copyright (C) 2024, Pelican Project, Morgridge Institute for Research
+*
+* Licensed under the Apache License, Version 2.0 (the "License"); you
+* may not use this file except in compliance with the License.  You may
+* obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*
+***************************************************************/
+
+package pelican_url
+
+import (
+	"fmt"
+	"net/url"
+
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+)
+
+// Available pelican URL query params
+const (
+	QueryRecursive    string = "recursive"
+	QueryPack         string = "pack"
+	QueryDirectRead   string = "directread"
+	QuerySkipStat     string = "skipstat"
+	QueryPreferCached string = "prefercached"
+
+	PackValueAuto  string = "auto"
+	PackValueTar   string = "tar"
+	PackValueTarGz string = "tar.gz"
+	PackValueTarXz string = "tar.xz"
+	PackValueZip   string = "zip"
+)
+
+func ParseQuery(query string) (PelicanURLValues, error) {
+	values, err := url.ParseQuery(query)
+	if err != nil {
+		return nil, err
+	}
+
+	return PelicanURLValues(values), nil
+}
+
+// Like url.Values.Get, this only gets the first value for a given key
+func (pvs PelicanURLValues) Get(key string) string {
+	if values, ok := pvs[key]; ok {
+		return values[0]
+	}
+	return ""
+}
+
+func (pvs PelicanURLValues) Add(key string, val string) {
```

Comment:
I might have missed this, but where in the code is 'Add' used?

Reply:
I don't recall if/where it's currently used, but it's something I've used enough with regular URL parsing that I wanted to be forward thinking.